ssl/statem/statem_lib.c

```diff
@@ -1038,6 +1037,203 @@ static int ssl_add_cert_chain(SSL_CONNECTION *s, WPACKET *pkt, CERT_PKEY *cpk, i
+    return NULL;
+}
+
+int tls_process_rpk(SSL_CONNECTION *sc, PACKET *pkt, EVP_PKEY **match)
+{
+    int ret = 0;
+    EVP_PKEY *pkey = NULL;
+    RAW_EXTENSION *rawexts = NULL;
+    PACKET extensions;
+    int i;
+    PACKET context;
+    unsigned long cert_list_len, spki_len;
+    const unsigned char *spki, *spkistart;
+    SSL_CTX *sctx = SSL_CONNECTION_GET_CTX(sc);
+
+    *match = NULL;
+    if (SSL_CONNECTION_IS_TLS13(sc)) {
+        if (!PACKET_get_length_prefixed_1(pkt, &context)) {
+            SSLfatal(sc, SSL_AD_DECODE_ERROR, SSL_R_INVALID_CONTEXT);
+            goto err;
+        }
+        if (sc->server) {
+            if (sc->pha_context == NULL) {
+                if (PACKET_remaining(&context) != 0) {
+                    SSLfatal(sc, SSL_AD_DECODE_ERROR, SSL_R_INVALID_CONTEXT);
+                    goto err;
+                }
+            } else {
+                if (!PACKET_equal(&context, sc->pha_context, sc->pha_context_len)) {
+                    SSLfatal(sc, SSL_AD_DECODE_ERROR, SSL_R_INVALID_CONTEXT);
+                    goto err;
+                }
+            }
+        } else {
+            if (PACKET_remaining(&context) != 0) {
+                SSLfatal(sc, SSL_AD_DECODE_ERROR, SSL_R_INVALID_CONTEXT);
+                goto err;
+            }
+        }
+    }
+
+    if (!PACKET_get_net_3(pkt, &cert_list_len)
+            || PACKET_remaining(pkt) != cert_list_len
+            || PACKET_remaining(pkt) == 0) {
+        SSLfatal(sc, SSL_AD_DECODE_ERROR, SSL_R_LENGTH_MISMATCH);
+        goto err;
+    }
+
+    /* RPK only admits a single CertificateEntry in the list, so no loop. */
+    if (!PACKET_get_net_3(pkt, &spki_len)
+            || spki_len == 0
+            || !PACKET_get_bytes(pkt, &spki, spki_len)) {
+        SSLfatal(sc, SSL_AD_DECODE_ERROR, SSL_R_LENGTH_MISMATCH);
+        goto err;
+    }
+    spkistart = spki;
+    if ((pkey = d2i_PUBKEY_ex(NULL, &spki, spki_len, sctx->libctx, sctx->propq)) == NULL
+            || spki != (spkistart + spki_len)) {
+        SSLfatal(sc, SSL_AD_DECODE_ERROR, SSL_R_LENGTH_MISMATCH);
+        goto err;
+    }
+    if (EVP_PKEY_missing_parameters(pkey)) {
+        SSLfatal(sc, SSL_AD_INTERNAL_ERROR,
+                 SSL_R_UNABLE_TO_FIND_PUBLIC_KEY_PARAMETERS);
+        goto err;
+    }
+    /* Verify the received key is a configured one. */
+    for (i = 0; i < sk_EVP_PKEY_num(sc->peer_rpks); i++) {
+        EVP_PKEY *conf_pkey = sk_EVP_PKEY_value(sc->peer_rpks, i);
+
+        /*
+         * EVP_PKEY_eq() will throw an error for different types,
+         * so, check that first, and then explicitly check for 1
+         */
+        if (EVP_PKEY_get_id(pkey) == EVP_PKEY_get_id(conf_pkey)
+                && EVP_PKEY_eq(conf_pkey, pkey) == 1) {
+            *match = conf_pkey;
+            break;
+        }
+    }
+    if (*match == NULL) {
+        SSLfatal(sc, SSL_AD_ILLEGAL_PARAMETER,
+                 SSL_R_INVALID_CERTIFICATE_OR_ALG);
```

Comment:
This should not be unconditional.  It should be possible to keep going even if the peer key is not matched (by overriding failure in a callback as in X.509).

Reply:
The idea is, if you're connecting to a peer and using RPK, then you better know what that RPK is, or it's _extremely_ insecure. That being said, the functionality does not mirror the certificate validation functionality.

Reply:
> The idea is, if you're connecting to a peer and using RPK, then you better know what that RPK is, or it's _extremely_ insecure. That being said, the functionality does not mirror the certificate validation functionality.

Well, it isn't that simple. Some applications continue the handshake on authentication failure, and then gracefully disconnect at the application layer, or negotiate RPK just to avoid sending useless certificates over the wire, but don't care what's in either the certificate or the RPK.  Just because RPK is enabled, does not actually mean you know the peer's expected keys.